mlflow/langchain/__init__.py

```diff
@@ -314,13 +329,66 @@ def log_model(
 
 
 def _save_model(model, path):
-    model.save(path)
+    import langchain
+
+    model_data_path = os.path.join(path, _MODEL_DATA_FILE_NAME)
+    model_data_kwargs = {_MODEL_DATA_KEY: _MODEL_DATA_FILE_NAME}
+
+    if type(model) == langchain.chains.llm.LLMChain:
+        model.save(model_data_path)
+    elif type(model) == langchain.agents.agent.AgentExecutor:
+        if model.agent and model.agent.llm_chain:
+            model.agent.llm_chain.save(model_data_path)
+
+        if model.agent:
+            agent_data_path = os.path.join(path, _AGENT_DATA_FILE_NAME)
+            model.save_agent(agent_data_path)
+            model_data_kwargs = {**model_data_kwargs, _AGENT_DATA_KEY: _AGENT_DATA_FILE_NAME}
+
+        if model.tools:
+            tools_data_path = os.path.join(path, _TOOLS_DATA_FILE_NAME)
+            with open(tools_data_path, "wb") as f:
+                cloudpickle.dump(model.tools, f)
```

Comment:
Does pickling the model tools work in most cases? Particularly, does it work with LangChain + OpenAI?

Reply:
Yes I did test it out with the example in the simple chain use case which uses OpenAI as a llm tool and it works :)